adapters/datadog-porting-guide.md

```diff
@@ -0,0 +1,101 @@
+# Porting Guide
+
+## Purpose
+
+Aid developers who wish to port existing datadog (dd-trace-rb) instrumentation adapters to opentelemetry.
+
+## Interface
+
+* OpenTelemetry `Adapter.install` -- does all integration work, which is custom depending on instrumented library
+
+## Basic structure
+
+* Add Gemfile, opentelemetry-adapters-#{name}.gemspec
+* Add runnable (docker) example (using its own Gemfile)
+```
+$ docker-compose run ex-adapter-myinstrumentation bundle install
+$ docker-compose run ex-adapter-myinstrumentation
+bash-5.0$ ruby trace_demonstration.rb
+```
+* Rakefile
+* `tests/test_helper.rb`
+* Integrate rubocop (see https://github.com/open-telemetry/opentelemetry-ruby/pull/172#pullrequestreview-349183775)
+
+## Examples and template adapters
+
+* `adapters/faraday` and `adapters/sinatra` were the earliest, simplest implementations
+
+## Implementation
+
+* It's ok to use `require_relative` for files that are internal to the project
+* `Span#status` can be set via helper `OpenTelemetry::Trace::Status.http_to_status`
+* Don't load integration implementation (require file) until `install` ('patch')-time
+* Most times, only want to run installation (via `#install`) once, but need to
+  consider being able to `reset` somehow for, e.g., testing
+
+### otel: tracer
+
+* `Tracer` `:name` and `:version` should come from instrumentation/adapter, not from the instrumented library
+
+### otel: span
+
+* Wrap instrumented-library operations in `span` (via `tracer.in_span` (block), or `tracer.start_span` (returns `Span`))
+  * `Span.finish` should be called if using `tracer.start_span`
+* Span `:with_parent` defaults to `current_span`
+
+### otel: span.name (dd-trace-rb: 'resource')
+
+* Prefer "low-cardinality" names (see https://github.com/open-telemetry/opentelemetry-specification/pull/416)
+
+### otel: span.attributes (dd-trace-rb: 'tags')
+
+* `Span` attribute keys should be strings, *not* symbols
+* `Span` attribute values should be strings, *not* symbols
+* Prefer to populate span attributes via method arguments (e.g., `tracer.in_span(attributes: ...`) instead of `span.set_attribute`
+* Some `Span` attribute naming is based on [semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md)
+* `Span` `:kind` defaults to `:internal` (other available options include `:client` or `:server`)
+
+### Runnable example
+* Gemfile should include "opentelemetry-adapters-#{adapter_name}"
+
+## Testing
+
```

Comment:
We should mention that when converting datadog instrumentation we should preserve all applicable tests from the Datadog version. Things that test datadog specific configuration, or things that are not being ported over can be omitted.